pkg/operator/controller/clientca-configmap/controller_test.go

```diff
@@ -0,0 +1,304 @@
+package clientcaconfigmap
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/openshift/cluster-ingress-operator/pkg/manifests"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// Test_Reconcile verifies that the controller's Reconcile method behaves
+// correctly.
+func Test_Reconcile(t *testing.T) {
+	exampleTime, err := time.Parse(time.DateTime, "2006-01-02 15:04:05")
+	if err != nil {
+		t.Fatal(err)
+	}
+	ic := func(deleted bool, configmapName string, additionalFinalizers ...string) *operatorv1.IngressController {
+		var ts *metav1.Time
+		if deleted {
+			ts = &metav1.Time{Time: exampleTime}
+		}
+		return &operatorv1.IngressController{
+			ObjectMeta: metav1.ObjectMeta{
+				DeletionTimestamp: ts,
+				Finalizers:        append([]string{manifests.IngressControllerFinalizer}, additionalFinalizers...),
+				Namespace:         "openshift-ingress-operator",
+				Name:              "test",
+			},
+			Spec: operatorv1.IngressControllerSpec{
+				ClientTLS: operatorv1.ClientTLS{
+					ClientCA: configv1.ConfigMapNameReference{
+						Name: configmapName,
+					},
+				},
+			},
+		}
+	}
+	cm := func(namespace, name string, data map[string]string) *corev1.ConfigMap {
+		return &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: namespace,
+				Name:      name,
+			},
+			Data: data,
+		}
+	}
+	var (
+		expectedData = map[string]string{
+			"ca-bundle.pem": "certificate",
+		}
+		unexpectedData = map[string]string{
+			"garbage": "garbage",
+		}
+	)
+	tests := []struct {
+		name            string
+		existingObjects []runtime.Object
+		expectCreate    []client.Object
+		expectUpdate    []client.Object
+		expectDelete    []client.Object
+	}{
+		{
+			name:            "ingresscontroller absent",
+			existingObjects: []runtime.Object{},
+			expectCreate:    []client.Object{},
+			expectUpdate:    []client.Object{},
+			expectDelete:    []client.Object{},
+		},
+		{
+			name: "no configmap specified",
+			existingObjects: []runtime.Object{
+				ic(false, ""),
+			},
+			expectCreate: []client.Object{},
+			expectUpdate: []client.Object{},
+			expectDelete: []client.Object{},
+		},
+		{
+			name: "source configmap absent, target configmap absent, finalizer absent",
```

Comment:
In this test case, the IngressController exists and specifies the "ca-bundle" configmap, but the configmap doesn't exist, so the controller can add the finalizer but cannot copy the configmap.

Reply:
If it's not too much trouble, it would be helpful to specify in the test `name` what the outcome of the test is supposed to be.  So here, because there is no `cm` in the `existingObjects`, there is no `cm` in the `expectUpdate`?  And, it is ok for the configmap name to be there in the `ic` even if there is no `cm` object? And we only add a finalizer if there is a configmap name, even if there is no matching `cm` object?

Reply:
> If it's not too much trouble, it would be helpful to specify in the test name what the outcome of the test is supposed to be. 

Done in https://github.com/openshift/cluster-ingress-operator/compare/3a3fb20502db4bb46db3de1ffef89faa5f84cc6a..e9ceaf5969c681a0ba773d5f97bc661703dfbf17.

> So here, because there is no cm in the existingObjects, there is no cm in the expectUpdate? 

Correct.  There is no source configmap, so the controller cannot make a copy of it.

> And, it is ok for the configmap name to be there in the ic even if there is no cm object? 

It can happen, so the controller needs to handle the case.  The router pods cannot start until the configmap exists though because the router pods have a volume mount for the configmap.  

> And we only add a finalizer if there is a configmap name, even if there is no matching cm object?

Correct.